### PR TITLE
Add broadcasting of weights to bce/bce_with_logits

### PR DESCRIPTION
**Pull request by alykhantejani**
This PR allows weights to `BCELoss` and `BCEWithLogitsLoss` to be the same shape as the target tensor, which is useful for tasks such as multi-label sample by sample weighting (as requested in #1792 [here](https://github.com/pytorch/pytorch/pull/1792#issuecomment-311132426)).

This takes advantage of broadcasting and we no longer need to expand the weights to match the target tensor. We also no longer need `torch/nn/_functional/thnn/loss.py` as the glue code for weight re-shaping happens in the functional interface now.

### CODE REVIEW

**Comment by soumith** [comment]
thanks Aly!
